src/impression.js

```diff
@@ -47,15 +53,72 @@ export function maybeTrackImpression(win) {
     // avoid duplicate tracking.
     win.location.hash = '';
   }
+
   viewer.whenFirstVisible().then(() => {
-    invoke(win, clickUrl);
+    // TODO(@zhouyx) We need a timeout here?
+    // TODO(@zhouyx) need test with a real response.
+    invoke(win, clickUrl).then(response => {
+      if (!response) {
+        return;
+      }
+      applyResponse(win, viewer, response);
+    });
   });
 }
 
+/**
+ * Send the url to ad server and wait for its response
+ * @param {!Window} win
+ * @param {string} clickUrl
+ * @return {!Promise}
+ */
 function invoke(win, clickUrl) {
-  xhrFor(win).fetchJson(clickUrl, {
+  if (getMode().localDev && !getMode().test) {
+    clickUrl = 'http://localhost:8000/impression-proxy?url=' + clickUrl;
+  }
+  return xhrFor(win).fetchJson(clickUrl, {
     credentials: 'include',
     requireAmpResponseSourceOrigin: true,
   });
   // TODO(@cramforce): Do something with the result.
 }
+
+/**
+ * parse the response back from ad server
+ * Set for analytics purposes
+ * @param {!Window} win
+ * @param {!Object} response
+ */
+function applyResponse(win, viewer, response) {
+  const adLocation = response['location'];
+
+  // If there's a tracking_url, need to redirect to that url.
+  const adTrackingUrl = response['tracking_url'];
+  if (adTrackingUrl) {
+    // TODO(@zhouyx) Confirm we assume the tracking_url is not
+    // a cdn.ampproject page, and we don't do viewer redirect
+    openWindowDialog(win, adTrackingUrl, '_top');
+    return;
+  }
+
+  // If adLocation is not an amp page, need to redirect to it.
+  if (adLocation && adLocation.indexOf('https://cdn.ampproject.org') != 0) {
+    // TODO(@zhouyx) Confirm we assume the tracking_url is not
+    // a cdn.ampproject page, and we don't do viewer redirect
+    openWindowDialog(win, adLocation, '_top');
+    return;
+  }
+
+  // If have gclid, or location contains gclid. add it to location hash
+  // Set gclid.
+  // If provided use it, if not try to find in adLocation from response
+  let gclid = response['gclid'];
+  if (!gclid) {
+    if (adLocation.indexOf('gclid=') > 0) {
+      gclid = adLocation.substr(adLocation.indexOf('gclid=') + 6);
```

Comment:
The code will go away, but please never parse a URL like that :) Use the utilities in `url.js`